provider/provider.go

```diff
@@ -70,6 +78,32 @@ func NewPinnedProvider(onlyRoots bool, pinning pin.Pinner, fetchConfig fetcher.F
 	}
 }
 
+// NewBufferedProvider returns a KeyChanFunc supplying keys from a given
+// KeyChanFunction, but buffering keys in memory if we can read them faster
+// they are consumed.  This allows the underlying KeyChanFunc to finish
+// listing pins as soon as possible releasing any resources, locks, at the
+// expense of memory usage.
+func NewBufferedProvider(pinsF KeyChanFunc) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		pins, err := pinsF(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		queue := chanqueue.New[cid.Cid]()
+
+		go func() {
+			in := queue.In()
+			defer close(in)
+			for c := range pins {
+				in <- c
+			}
+		}()
+
+		return queue.Out(), nil
```

Comment:
If you want to wait for me to make a small change in chanqueue to allow it to accept a read-only input channel, then  the code can be simplified to this suggestion:
```suggestion
		queue := chanqueue.New(chanqueue.WithInputRdOnly[cid.Cid](pins))
		return queue.Out(), nil
```

Reply:
It will operate the same, but without the additional goroutine. `queue.Out()` will be closed when `pins` in closed and items are read.

The only difference from your original code is that canceling the context does not cause the output channel to immediately close, requiring all the output to be read first. If handling the closed context immediately is a concern, then it can be done by:
```go
context.AfterFunc(ctx, func() {
	queue.Shutdown()
})
```
... as Shutdown drains the Chanqueue.

Reply:
Perfect thanks!